cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,84 @@ cuML is thread safe so its functions can be called from multiple host threads if
 
 The implementation of cuML is single threaded.
 
+## Public cuML interface
+### Motivation
+Public cuML interface is stateless for two main reasons:
+1. Easily pickle ML algo's state information (model, hyper-params, etc), at the python layer
+2. Provide a proper C-binding to interface with languages that don't understand C++
+Thus, this section lays out guidelines for managing state along the API of cuML.
+
+### General guideline
+As mentioned before, functions exposed via the base cuML-c++ layer must be stateless. Things that are OK to be expoesd on the interface:
+1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure) - one can use [std::is_pod](https://en.cppreference.com/w/cpp/types/is_pod) in C++11 to check POD types.
+2. `cumlHandle` - since it stores GPU-related state which has nothing to do with the model/algo state. If you're working on a C-binding, use `cumlHandle_t`([reference](src/cuML_api.h)), instead.
+3. Pointers to POD types (explicitly putting it out, even though can be considered as a POD).
+Internally, for the C++ base layer atleast, these stateless functions are free to use their own temporary classes, as long as they are not exposed on the interface.
+
+### `libcuml++.so` (aka C++ API)
+Taking decisiontree-classifier algo as an example, the following way of exposing its API would be wrong according to the guidelines in this section. Because, this API exposes a non-POD C++ class object along the C++ API.
```

Comment:
```suggestion
Using the Decision Tree Classifier algorithm as an example, the following way of exposing its API would be wrong according to the guidelines in this section, since it exposes a non-POD C++ class object in the C++ API:
```